notes.txt

```diff
@@ -0,0 +1,241 @@
+Given this reality,
+1) I think the naive approach is to track user's unbondings in stakeibc module state, then transfer 
+the unbonded tokens from the delegation account to the user's account as soon as the unbonding period 
+ends. Staking module's unbonding-delegations object has completion_time: "2022-06-12T23:10:51.070726210Z" 
+and creation_height: "390" fields that help us time this correctly -- we need to make sure we transfer 
+the unbonded tokens rather than sweep them back to staking
+
+2) A slightly better approach may be to sweep only the sum of amounts liquid staked since the last epoch 
+(by summing up deposit receipts), rather than staking the entire delegation account balance. This would 
+alleviate the need to transfer the funds out before delegation account token are re-staked (because the 
+unbonded tokens would remain in the account until transfered to the user).
+
+3) Another approach could be to track unbonded amounts each in a FIFO queue in stakeibc module state, 
+and, at each epoch, while the balance in the delegation account is  >= the amount of the top item in 
+the queue, transfer funds to that user's account and pop the queue. If the queue is emptied, stake the 
+remaining funds. Haven't thought this one through fully
+
+I vote for (2) for its simplicity, but (3) could be interesting down the line bc it allows 
+unbondings in less than 21 days
+
+# Records we need to keep
+
+
+
+
+*******************************************************************
+******************************DEPOSITS*****************************
+*******************************************************************
+
+# SECTION 1
+# 0 Update DepositRecord type
+  UPDATE DEPOSIT RECORD
+  message DepositRecord {
+    uint64 id = 1; // epoch_num:chain_id
+    int64 amount = 2; // total amount
+    string denom = 3; // tokens
+    string hostZoneId = 4; // can remove - fetch with denom
+    // the next thing we should do
+    enum RecordType {
+      DEPOSIT = 0;
+      REWARDS = 1;
+      WITHDRAWAL = 2;
+    }
+    enum Status {
+      TRANSFER = 0;
+      STAKE = 1;
+    }
+    Status status = 5; 
+  }
+  - Add next status types
+      - transfer: the balance must be transferred to the host zone
+      - stake: the balance must be staked
+# 1 Aggregate into a single deposit record / successful epoch / chain
+# 2 Update DepositRecord type
+# 3 Update Epoch processing code
+
+>>> EPOCH START
+----------
+
+# users call liquid stake
+def liquidStake(tokens)
+  currentRecord = getRecord(epochId)
+  currentRecord.amount += tokens.amount
+
+>>> EPOCH END
+for record in depositRecords:
+  hz = getHz(record)
+  if record.status == TRANSFER:
+    def ibcTransferAck(ack):
+      if ack == success:
+        record = getRecord(ack)
+        delete(record)
+    # add record deletion logic to ack
+    ibcTransfer.ackFunction = ibcTransferAck
+    # ibc transfer the tokens over
+    ibcTransfer(to=hz.depositAccount(), amount=record.amount)
```

Comment:
Logic makes sense, good thinking to use the ack to delete records. I thought we were planning to batch transfers from the module account to the delegation account?